stock_cycle_count/models/stock_cycle_count_rule.py

```diff
@@ -0,0 +1,236 @@
+                      'per period.')
+                )
+            if rec.periodic_count_period < 0:
+                raise ValidationError(
+                    _('You cannot define a negative period.')
+                )
+
+    @api.onchange('location_ids')
+    def _get_warehouses(self):
+        """Get the warehouses for the selected locations."""
+        wh_ids = []
+        for loc in self.location_ids:
+            wh_ids.append(loc.get_warehouse().id)
+        wh_ids = list(set(wh_ids))
+        self.warehouse_ids = self.env['stock.warehouse'].browse(wh_ids)
+
+    name = fields.Char(required=True)
+    rule_type = fields.Selection(
+        selection="_selection_rule_types",
+        string='Type of rule', required=True,
+    )
+    rule_description = fields.Char(
+        string='Rule Description', compute='_compute_rule_description',
+    )
+    active = fields.Boolean(default=True)
+    periodic_qty_per_period = fields.Integer(
+        string='Counts per period', default=1,
+    )
+    periodic_count_period = fields.Integer(string='Period in days')
+    turnover_inventory_value_threshold = fields.Float(
+        string='Turnover Inventory Value Threshold',
+    )
+    currency_id = fields.Many2one(
+        comodel_name='res.currency', string='Currency',
+        compute='_compute_currency_id',
+    )
+    accuracy_threshold = fields.Float(
+        string='Minimum Accuracy Threshold', digits=(3, 2),
+    )
+    apply_in = fields.Selection(
+        string='Apply this rule in:',
+        selection=[('warehouse', 'Selected warehouses'),
+                   ('location', 'Selected Location Zones.')],
+        default='warehouse',
+    )
+    warehouse_ids = fields.Many2many(
+        comodel_name='stock.warehouse',
+        relation='warehouse_cycle_count_rule_rel', column1='rule_id',
+        column2='warehouse_id', string='Warehouses where applied',
+    )
+    location_ids = fields.Many2many(
+        comodel_name='stock.location',
+        relation='location_cycle_count_rule_rel', column1='rule_id',
+        column2='location_id', string='Zones where applied',
+    )
+
+    def compute_rule(self, locs):
+        if self.rule_type == 'periodic':
+            proposed_cycle_counts = self._compute_rule_periodic(locs)
+        elif self.rule_type == 'turnover':
+            proposed_cycle_counts = self._compute_rule_turnover(locs)
+        elif self.rule_type == 'accuracy':
+            proposed_cycle_counts = self._compute_rule_accuracy(locs)
+        return proposed_cycle_counts
+
+    @api.model
+    def _propose_cycle_count(self, date, location):
+        cycle_count = {
+            'date': date.strftime(DEFAULT_SERVER_DATETIME_FORMAT),
+            'location': location,
+            'rule_type': self
+        }
+        return cycle_count
+
+    @api.model
+    def _compute_rule_periodic(self, locs):
+        cycle_counts = []
+        for loc in locs:
+            last_inventories = self.env['stock.inventory'].search([
```

Comment:
Maybe the search can be directly order by date desc and limit=1 to fetch less records.

Reply:
right, done.